supervise.go

```diff
@@ -539,13 +538,15 @@ var services struct {
 	S []*service
 }
 
-// killSupervisedServices is called before rebooting when upgrading, allowing
-// processes to terminate in an orderly fashion.
-func killSupervisedServices() {
+// signalSupervisedServices sends a given signal to all non-stopped processes.
+// It returns the corresponding processState to allow waiting for a given state.
+func signalSupervisedServices(signal syscall.Signal) []*processState {
 	services.Lock()
 	defer services.Unlock()
+
+	states := make([]*processState, 0, len(services.S))
 	for _, s := range services.S {
-		if s.Stopped() {
+		if s.Stopped() && s.state.Get() == Stopped {
```

Comment:
Can you add a comment explaining why checking Stopped() is not enough in this case?

Reply:
Added

```
		// s.Stopped() only checks the "stopped" flag of the service (if it shouldn't restart).
		// We check the actual state as well to be sure to re-send a signal if we are already
		// in the "Stopping" state.
```